arch/xtensa/src/esp32/esp32_qencoder.c

```diff
@@ -0,0 +1,928 @@
+  uint32_t  ch1_pulse_sig; /* ch1 pulse signal index */
+  uint32_t  ch1_ctrl_sig;  /* ch1 ctrl signal index */
+  uint16_t  filter_thres;  /* Filter threshold for this PCNT Unit */
+};
+
+/* NOTE: we are using Quadrature Encoder in X4 mode on ESP32 PCNT, then
+ * instead of using 'pulse_gpio' and 'ctrl_gpio' names, we only use ch0_gpio
+ * and ch1_gpio names. It avoid confusion, since the same signal that is used
+ * on pin 'pulse' of CH0 is also connected to 'ctrl' pin of the CH1 and
+ * 'ctrl' pin of CH0 is also connected on 'pulse' pin of CH1.
+ */
+
+/* Overall, RAM-based state structure */
+
+struct esp32_lowerhalf_s
+{
+  /* The first field of this state structure must be a pointer to the
+   * lower-half callback structure:
+   */
+
+  const struct qe_ops_s *ops; /* Lower half callback structure */
+
+  /* ESP32 driver-specific fields: */
+
+  const struct esp32_qeconfig_s *config; /* static configuration */
+
+  bool inuse; /* True: The lower-half driver is in-use */
+
+  volatile int32_t position; /* The current position offset */
+
+  spinlock_t lock; /* Device specific lock. */
+};
+
+/****************************************************************************
+ * Private Function Prototypes
+ ****************************************************************************/
+
+/* Helper functions */
+
+#if defined(CONFIG_DEBUG_SENSORS) && defined(CONFIG_DEBUG_INFO)
+static void esp32_dumpregs(struct esp32_lowerhalf_s *priv,
+                           const char *msg);
+#else
+#  define esp32_dumpregs(priv,msg)
+#endif
+
+static struct esp32_lowerhalf_s *esp32_pcnt2lower(int pcnt);
+
+/* Interrupt handling */
+
+#if 0 /* FIXME: To be implement */
+static int esp32_interrupt(int irq, void *context, void *arg);
+#endif
+
+/* Lower-half Quadrature Encoder Driver Methods */
+
+static int esp32_setup(struct qe_lowerhalf_s *lower);
+static int esp32_shutdown(struct qe_lowerhalf_s *lower);
+static int esp32_position(struct qe_lowerhalf_s *lower,
+                          int32_t *pos);
+static int esp32_setposmax(struct qe_lowerhalf_s *lower, uint32_t pos);
+static int esp32_reset(struct qe_lowerhalf_s *lower);
+static int esp32_setindex(struct qe_lowerhalf_s *lower, uint32_t pos);
+static int esp32_ioctl(struct qe_lowerhalf_s *lower, int cmd,
+                       unsigned long arg);
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+/* The lower half callback structure */
+
+static const struct qe_ops_s g_qecallbacks =
+{
+  .setup     = esp32_setup,
+  .shutdown  = esp32_shutdown,
+  .position  = esp32_position,
+  .setposmax = esp32_setposmax,
+  .reset     = esp32_reset,
+  .setindex  = esp32_setindex,
+  .ioctl     = esp32_ioctl,
+};
+
+/* Per-pcnt state structures */
+
+#ifdef CONFIG_ESP32_PCNT_U0_QE
+static const struct esp32_qeconfig_s g_pcnt0config =
+{
+  .pcntid        = 0,
```

Comment:
Description in struct is `uint8_t   pcntid;        /* PCNT ID {1,2,3,4,5,6,7,8} */`, so this assignment is confusing

Reply:
The comment is incorrect, I will fix it, thank you!